Use JSpecify samples as test suite

wmdietl opened the pull request.

cpovirk reviewed (commented):
Thanks!

I pushed a couple commits.

Note that the version-bump commit will require you to pull the latest changes into your corresponding clone of our CF fork and then rebuild it.

The other commit eliminates a few differences between the actual and expected results.

The remaining failures (2 of them, though each occurs in both strict mode and lenient mode, so it counts as 4):

- `fullyQualified` fails because (I think) the test driver tries to run solely on the directory `fullyQualified/fullyqualified/caller` instead of running on all of `fullyQualified` at once.
- `Annotated*` and its `NotNullAware*` variants fail. I see that this is another case of multiple failures on one line, as you'd referred to in another conversation. I can imagine multiple ways forward:
    - As long as we recognize one error on a line, ignore the others.
    - Edit our checker and/or CF itself to _really_ ignore annotations on wildcards and type parameters, not to just report an error about them but then still respect them.
    - (I was going to also suggest declaring bounds that make the declarations legal under CF's interpretation of annotations on wildcards and type parameters. But it looks like I explicitly wanted to test both bounded and unbounded cases, since they are sometimes represented differently.)
    - possibly other solutions I'm not thinking of

cpovirk commented:
> The remaining failures (2 of them, though each occurs in both strict mode and lenient mode, so it counts as 4):

Sorry, I didn't say the important part: That's the number of failures _relative to what I expect_. There are indeed additional failures when running directly against the current samples. As discussed, we'll want to identify those expected failures to this test code in some way.

cpovirk commented:
The results I expect: https://github.com/cpovirk/jspecify/commit/944271f44b33e54001014fbf8e487a92cde1a07f

(Note that I was not careful about which specific token I applied: If I got an error in strict mode, I added jspecify_nullness_mismatch, and if I got an error in lenient mode, I added jspecify_nullness_not_enough_information. It's *possible* that some of those "should" be another token, like jspecify_nullness_intrinsically_not_nullable or something. But they're probably right, and they should be close enough for our purposes.)

wmdietl commented:
I made a few more changes:
- split off an included minimal test from the JSpecify samples, to make it easier to see what failed (this could be improved by making it a separate workflow step, if you think that's better) (In the future we could also add more fine-grained checks here that go beyond what JSpecify mandates.)
- simplified the debug output a bit.

Have a look at the GitHub checks output with the last commit.
How about integrating this PR and then improving things further?

I'm not sure how to read your expected results commit. Are these changes that you want to integrate in jspecify samples? Or regressions that this checker has?

In your description, you had: 'If I got an error in strict mode, I added jspecify_nullness_mismatch, and if I got an error in lenient mode, I added jspecify_nullness_not_enough_information.'
Should this be the other way around? If you get something in lenient mode, it is definitely a `jspecify_nullness_mismatch`. The errors in strict mode are the additional `jspecify_nullness_not_enough_information` errors.

wmdietl commented on build.gradle:
```diff
@@ -14,34 +17,43 @@ configurations {
 ext {
     parentDir = file("${rootDir}/../").absolutePath
     cfHome = "${parentDir}/checker-framework"
+    jspecifyHome = "${parentDir}/jspecify"
```
I think that shouldn't be hard to do. I'll work on that once these build changes are in.

cpovirk commented:
Sorry, I let myself get sidetracked a bunch this morning.

I'll merge this now and then respond to the remaining discussions.

cpovirk commented:
- The new GitHub Actions output looks cleaner. Thanks!
- The `Annotated*` and its `NotNullAware*` variants are fixed by https://github.com/jspecify/checker-framework/commit/1db27bac27507a46a9cc7a894df5e9c8d64c2722 and https://github.com/jspecify/checker-framework/commit/235b632812f838e9a9188cf5af842ae4e401e55f. (Possibly I should have fixed the problem in our checker instead: https://github.com/jspecify/checker-framework/issues/13)
- `fullyQualified` is still failing because of the multiple-package problem. It's just now occurring to me that there is no actual need for us to have 2 packages at all: We can fully qualify the names of classes even if they're in the same package. (It's possible that there is some value to having that Java package be "nested" inside another Java package, but we can still do that.) I will take a stab at this.

> I'm not sure how to read your expected results commit. Are these changes that you want to integrate in jspecify samples? Or regressions that this checker has?

Sorry for giving you a puzzle to work through. That commit shows the current, incorrect output of the checker according to my hacky test script. So I wouldn't want to modify the jspecify sample files in place. What I _might_ do is create a branch in the jspecify repo that contains those changes. Then we could run our tests against that. Or, better yet, I would keep the existing comments in place but modify them to include additional information about where our checker deviates. Internally, I have extra tokens -- "cf_no_error"/"cf_no_warning" and "cf_produces_warning"/"cf_produces_error" -- for cases in which the checker fails to report a problem or reports one spuriously. Do you know how hard it would be to tell your test setup things like "Look for an error on any line with jspecify_nullness_mismatch _except_ if it also contains cf_no_error?"

> In your description, you had: 'If I got an error in strict mode, I added jspecify_nullness_mismatch, and if I got an error in lenient mode, I added jspecify_nullness_not_enough_information.'
> Should this be the other way around? If you get something in lenient mode, it is definitely a `jspecify_nullness_mismatch`. The errors in strict mode are the additional `jspecify_nullness_not_enough_information` errors.

Yes, sorry again. I said it backward.

cpovirk commented:
> * `fullyQualified` is still failing because of the multiple-package problem. It's just now occurring to me that there is no actual need for us to have 2 packages at all: We can fully qualify the names of classes even if they're in the same package. (It's possible that there is some value to having that Java package be "nested" inside another Java package, but we can still do that.) I will take a stab at this.

Not quite fixed by https://github.com/jspecify/jspecify/commit/147409aaa5308f76bac2737164d9f6dfee26a655, then actually fixed by https://github.com/jspecify/jspecify/commit/2e5798ba1578efb971f151a60a4e3605eafae265.

cpovirk commented:
Also: I added some docs about this in https://github.com/jspecify/nullness-checker-for-checker-framework/commit/13a9595db4af20ba3e086f751da00a96e2cc7701 -- including a warning that my https://github.com/jspecify/nullness-checker-for-checker-framework/commit/2b4626c17b44b16577801f23fc26b456ca51ea77 makes `./gradlew jspecifySamplesTest` rebuild the Checker Framework unless you exclude that task with `-x ensureCheckerFrameworkBuilt`. We might have discussed whether to _always_ rebuild the Checker Framework -- even if it already exists: https://github.com/jspecify/nullness-checker-for-checker-framework/blob/13a9595db4af20ba3e086f751da00a96e2cc7701/build.gradle#L106 -- but I forget offhand.
